### ISSUE TITLE
关于distillation_loss中F.log_softmax和F.softmax不一致的问题

### ISSUE THREAD

**Issue by zhujiesuper** [mepeichun/Efficient-Neural-Network-Bilibili#1]
def distillation(y, labels, teacher_scores, temp, alpha):
    print("teacher_scores",teacher_scores)
    return nn.KLDivLoss()(F.log_softmax(y / temp, dim=1), F.softmax(teacher_scores / temp, dim=1)) * (
            temp * temp * 2.0 * alpha) + F.cross_entropy(y, labels) * (1. - alpha)

在次代码中，为啥F.log_softmax(y / temp, dim=1，用的log，而F.softmax(teacher_scores / temp, dim=1)不用log

**Comment by mepeichun** [comment]
手动实现一个KL divergence
```
import torch.nn.functional as F

# input_logit: [batch, class_num]
# target_logit: [batch, class_num]
def kl_loss(input_logit, target_logit):
    prob = F.softmax(target_logit, dim=-1)
    kl = torch.sum(prob  * (F.log_softmax(target_logit, dim=-1)
                                  - F.log_softmax(input_logit, dim=-1)), 1)
    return torch.mean(kl)
```

或者可以这样

```
import torch.nn.functional as F

# input_log_sofmax: [batch, 1]
# target_sofmax: [batch, 1]
def kl_loss(input_log_sofmax, target_sofmax):
    prob = target_sofmax
    kl = torch.sum(prob  * (input_log_sofmax - torch.log(target_sofmax)), 1)
    return torch.mean(kl)
```

很明显，pytorch采纳了第二种方案。避免了重复计算。

如果传入的参数为(input_log_sofmax, target_log_sofmax)，则需要将target_log_sofmax转化为target_sofmax计算概率

如果传入的参数为(input_sofmax, target_sofmax)，则需要将input_sofmax转化为input_log_sofmax

具体细节请参考 [https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html](https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html)

**Comment by zhujiesuper** [comment]
感谢，清晰易懂！太棒了




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "mepeichun/Efficient-Neural-Network-Bilibili"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年9月11日(星期五) 中午1:06
收件人:&nbsp;"mepeichun/Efficient-Neural-Network-Bilibili"<Efficient-Neural-Network-Bilibili@noreply.github.com&gt;;
抄送:&nbsp;"朱一大爱"<653725215@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [mepeichun/Efficient-Neural-Network-Bilibili] 提问！ (#1)





 
手动实现一个KL divergence
 import torch.nn.functional as F # input_logit: [batch, class_num] # target_logit: [batch, class_num] def kl_loss(input_logit, target_logit):     prob = F.softmax(target_logit, dim=-1)     kl = torch.sum(prob  * (F.log_softmax(target_logit, dim=-1)                                   - F.log_softmax(input_logit, dim=-1)), 1)     return torch.mean(kl)  
或者可以这样
 import torch.nn.functional as F # input_log_sofmax: [batch, class_num] # target_sofmax: [batch, class_num] def kl_loss(input_log_sofmax, target_sofmax):     prob = target_sofmax     kl = torch.sum(prob  * (input_log_sofmax - torch.log(target_sofmax)), 1)     return torch.mean(kl)  
很明显，pytorch推荐使用第二种方案。
 具体细节请参考 https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html

**Comment by Duducoco** [comment]
```
kl = torch.sum(prob  * (input_log_sofmax - torch.log(target_sofmax)), 1)
```
看了一个文章，感觉这一行代码应该改成下面的样子，不知道理解的对不对
[loss函数之KLDivLoss](https://blog.csdn.net/ltochange/article/details/118300003)
```
kl = torch.sum(prob  * (torch.log(target_sofmax)-input_log_sofmax), 1)
```

**Comment by JevenM** [comment]
> 

对的
